.github/workflows/ci.yml

```diff
@@ -135,3 +136,27 @@ jobs:
 
       # Check that the release automation works as expected
       - run: scripts/release/test.sh
+
+  # This check is allowed to fail, while Marker is still unstable. It's mostly
+  # a sanity check, to make sure we don't have any unintentional API breakage :)
+  semver-checks:
+    runs-on: ubuntu-latest
+    if: ${{ github.event.pull_request }}
```

Comment:
Let's add [`continue-on-error`](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idcontinue-on-error), so that if this job fails the workflow will still be displayed as green in the Actions tab of the repo.
```suggestion
    runs-on: ubuntu-latest
    # This CI check isn't required to pass just yet.
    # Marker isn't stable so semver breaking changes are allowed.
    continue-on-error: true
    if: ${{ github.event.pull_request }}
```

Reply:
This will require us to always look at the log, to see if the semver-check actually fails. :thinking:

For a second, I contemplated if it would be cool to automatically add/remove the `I-api-break` label. But then I considered the implications of giving PR edit permissions to a normal user editable CI job, and decided against it...

Reply:
I think having the check green by default is better than having it red :+1: 

Do you have another idea, how we could maybe get notified about the API breakage, without checking the job every time?

I considered checking for the `I-api-break` label and error if that one isn't present. However, the CI would not rerun if the label is added afterwards, resulting in a red CI again :/

The last test commit, contains some API breakage, just for a reference how it would look like.

Reply:
I wouldn't put the `continue-on-error` on the `cargo semver-checks` step itself.
This way the check itself will always be green, which I suppose isn't what we want. It should be read if a major breaking change was detected.

The `continue-on-error` at the job level makes it so that the workflow itself is displayed as green in the github action runs like here: https://github.com/rust-marker/marker/actions
<img width="375" alt="image" src="https://github.com/rust-marker/marker/assets/36276403/5b760927-1265-4611-82ea-9a4f18b770d5">

But inside the PR view the job should be displayed in red.

---

Regarding the notification. Yes! Auto-labeling the PR without granting the downstream fork broad write permissions is doable. Checkout out this [blog post](https://securitylab.github.com/research/github-actions-preventing-pwn-requests/).

I think we can use the [`workflow_run`](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#workflow_run) approach where the untrusted regular CI on `pull_request` event runs the semver checks, and in case the check detects a breaking change the CI job should mark itself as failed. Then, a followup privileged CI is triggered on `workflow_run` event. It will run in the context of the upstream repo (and not PR merge commit), so it will not observe any PR code modifications, and thus is safe. In our case it doesn't even need to check out the PR code or even do any code checkout whatsoever.

The unprivilleged `semver-checks` job needs to upload an artifact with the results of the checks, and the privileged workflow downloads the artifact ([docs](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#using-data-from-the-triggering-workflow)) and make take advantage from repo write permissions.

In our case the artifact should store the PR number and the boolean result of the check so that the workflow could determine if it needs to add a PR label (if one isn't already there), or remove it (if there is one).

Reply:
Having automated GitHub annotations would be the dream! I've subscribed to the state of the issue now, but it's likely that that solution will take its time.

I've now moved the [continue-on-error](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idcontinue-on-error), what I dislike a bit, is that it marks the PR as "failed". The CI job itself was less my concern.

I think the split CI approach, with one privileged CI for applying the label, would work well for the meantime. I'll try to find some time to test this.

Thank you for the suggestion!